src/autora/experimentalist/grid_.py

```diff
@@ -0,0 +1,124 @@
+"""Tools to make grids of experimental conditions."""
+from functools import singledispatch
+from itertools import product
+from typing import Sequence
+
+import pandas as pd
+
+from autora.state.delta import Result, State, wrap_to_use_state
+from autora.variable import Variable, VariableCollection
+
+
+@singledispatch
+def grid_pool(s, **kwargs):
+    """Function to create a sequence of conditions sampled from a grid of independent variables."""
+    raise NotImplementedError(
+        "grid_pool doesn't have an implementation for %s (type=%s)" % (s, type(s))
+    )
+
+
+@grid_pool.register(State)
+def grid_pool_on_state(s: State) -> State:
+    return wrap_to_use_state(grid_pool_from_variables)(s)
+
+
+@grid_pool.register(list)
+@grid_pool.register(tuple)
+def grid_pool_from_ivs(ivs: Sequence[Variable]) -> product:
+    """Creates exhaustive pool from discrete values using a Cartesian product of sets"""
+    # Get allowed values for each IV
+    l_iv_values = []
+    for iv in ivs:
+        assert iv.allowed_values is not None, (
+            f"gridsearch_pool only supports independent variables with discrete allowed values, "
+            f"but allowed_values is None on {iv=} "
+        )
+        l_iv_values.append(iv.allowed_values)
+
+    # Return Cartesian product of all IV values
+    return product(*l_iv_values)
```

Comment:
Hmm.. I'm not sure how I feel about this output. I think it is much more intuitive if the pooler would return a numpy array or pandas data frame.